pkg/stanza/operator/parser/container/parser.go

```diff
@@ -0,0 +1,359 @@
+}
+
+// crioConsumer receives log entries from the crioLogEmitter and
+// writes them to the output of the main parser
+func (p *Parser) crioConsumer(ctx context.Context) {
+	entriesChan := p.crioLogEmitter.OutChannel()
+	p.crioConsumers.Add(1)
+	defer p.crioConsumers.Done()
+	for entries := range entriesChan {
+		for _, e := range entries {
+			p.Write(ctx, e)
+		}
+	}
+}
+
+// Stop ensures that the internal recombineParser, the internal crioLogEmitter and
+// the crioConsumer are stopped in the proper order without being affected by
+// any possible race conditions
+func (p *Parser) Stop() error {
+	if !p.asyncConsumerStarted {
+		// nothing is started return
+		return nil
+	}
+
+	var stopErrs []error
+	err := p.recombineParser.Stop()
+	if err != nil {
+		stopErrs = append(stopErrs, fmt.Errorf("unable to stop the internal recombine operator: %w", err))
+	}
+	// the recombineParser will call the Process of the crioLogEmitter synchronously so the entries will be first
+	// written to the channel before the Stop of the recombineParser returns. Then since the crioLogEmitter handles
+	// the entries synchronously it is safe to call its Stop.
+	// After crioLogEmitter is stopped the crioConsumer will consume the remaining messages and return.
+	err = p.crioLogEmitter.Stop()
+	if err != nil {
+		stopErrs = append(stopErrs, fmt.Errorf("unable to stop the internal LogEmitter: %w", err))
+	}
+	p.crioConsumers.Wait()
+	return errors.Join(stopErrs...)
+}
+
+// detectFormat will detect the container log format
+func (p *Parser) detectFormat(e *entry.Entry) (string, error) {
+	value, ok := e.Get(p.ParseFrom)
+	if !ok {
+		return "", fmt.Errorf("entry cannot be parsed as container logs")
+	}
+
+	var raw string
+	switch m := value.(type) {
+	case string:
+		raw = m
+	default:
+		return "", fmt.Errorf("type '%T' cannot be parsed as container logs", value)
+	}
+
+	switch {
+	case dockerMatcher.MatchString(raw):
+		return dockerFormat, nil
+	case crioMatcher.MatchString(raw):
+		return crioFormat, nil
+	case containerdMatcher.MatchString(raw):
+		return containerdFormat, nil
+	}
+	return "", fmt.Errorf("entry cannot be parsed as container logs: %v", value)
+}
+
+// parseCRIO will parse a crio log value based on a fixed regexp
+func (p *Parser) parseCRIO(value any) (any, error) {
+	// Process can be called in parallel and in that case we want the parsing to happen
+	// once at a time to avoid changing each other's TimeParser
+	p.parserLock.Lock()
+	defer p.parserLock.Unlock()
+
+	var raw string
+	switch m := value.(type) {
+	case string:
+		raw = m
+	default:
+		return nil, fmt.Errorf("type '%T' cannot be parsed as crio container logs", value)
+	}
+
+	parseFromField := entry.NewAttributeField("time")
+	p.ParserOperator.TimeParser = &helper.TimeParser{
+		ParseFrom:  &parseFromField,
+		LayoutType: helper.GotimeKey,
+		Layout:     "2006-01-02T15:04:05.999999999Z07:00",
+	}
+	if err := p.ParserOperator.TimeParser.Validate(); err != nil {
+
+		return nil, err
+	}
```

Comment:
Well I originally thought that using the built-in TimeParser would be more straightforward but it seems that involves more complexity.
We can replace it with using the `timeutils` directly, yes. If I understood your suggestion correctly it is sth like 
f5ab018?

Reply:
Yeah, I would definitely use `timeutils` over the raw `time.Parse`, since there are many nuances which we've worked through in `timeutils`.

Your change is more or less what I had in mind, but we should also move the `timeLayout, err := timeutils.StrptimeToGotime("%Y-%m-%dT%H:%M:%S.%LZ")` calls out of the main processing path, since it's always going to yield the same result. (I'd also support adding a new `timeutils.MustStrptimeToGotime` since the error will be always or never.)